pkg/util/encoding/binary.go

```diff
@@ -0,0 +1,43 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package encoding is for utilities relating to the encoding package from the stdlib
+package encoding
+
+import (
+	"encoding"
+)
+
+// BinaryUnmarshalCallback returns a function that will decode the argument byte slice into *T
+// using `newFn` to create an instance of *T and the encoding.BinaryUnmarshaler interface to do the actual conversion.
+// `callback` will be called with the resulting *T.
+// If the argument byte slice is empty, callback will be called with `nil`.
+// Unmarshalling errors will be provided to the callback as the second argument. The data argument to the callback
+// may still be non-nil even if there was an error. This allows the callback to handle the allocated object, even
+// in the face of errors.
+// This function panics if `*T` does not implement encoding.BinaryUnmarshaler.
+func BinaryUnmarshalCallback[T any](newFn func() *T, callback func(*T, error)) func(buf []byte) {
+	// we use `any` as the type constraint rather than encoding.BinaryUnmarshaler because we are not allowed to
+	// callback with `nil` in the latter case. There is a workaround, but it requires specifying two type constraints.
+	// For sake of cleanliness, we resort to a runtime check here.
+	if _, ok := any(new(T)).(encoding.BinaryUnmarshaler); !ok {
+		panic("pointer type *T must implement encoding.BinaryUnmarshaler")
```

Comment:
Why do we allow panics in the code?

An alternative can be
```go
func BinaryUnmarshalCallback2(newFn func() encoding.BinaryUnmarshaler, callback func(encoding.BinaryUnmarshaler, error)) func(buf []byte) {
	return func(buf []byte) {
		if len(buf) == 0 {
			callback(nil, nil)
			return
		}

		d := newFn()
		if err := d.UnmarshalBinary(buf); err != nil {
			// pass d here so callback can choose how to deal with the data
			callback(d, err)
			return
		}
		callback(d, nil)
	}
}
```

Reply:
> Why do we allow panics in the code?

Panics are perfectly acceptable for programmer error (similar to asserts in other languages). See out of bounds access for slices, etc.

I cannot use your alternative for several reasons:
1. The reason listed in the comment in the code
2. I don't want to force the caller to cast from `encoding.BinaryUnmarshaler` to their type on each usage.
3. I want to ensure this function returns a pointer

Reply:
A single allocation when setting up the callback, not on every call. Please see earlier comments why constraints will not work.

Reply:
I have moved to a simpler implementation without the panic, trusting the user of this function has implemented the `encoding.BinaryUnmarshaler` interface correctly.